readme.md
<!--
  Created at: 06/09/2021 12:05:16 Wednesday
  Modified at: 09/17/2021 12:46:24 AM Friday

        Copyright (C) 2021 Thiago Navarro
  See file "license" for details about copyright
-->

<!-- Hello! Thanks for interest in my lib! :) -->

# Whois.nim

The Whois.nim is a simple whois client.
**With cache (kashae)!**

---

## Requirements

- The Nim ([Official website](https://nim-lang.org/)) programming language (Min version 1.4.0)
- Nimble ([Github](https://github.com/nim-lang/nimble)), the Nim's package manager

Hint: Install Nim with Choosenim ([Github](https://github.com/dom96/choosenim))

## Installation

Directly with Nimble

```bash
# Install directly with Nimble
nimble install whois
```

or

Directly with Nimble (with url)

```bash
# Install directly with Nimble (with url)
nimble install https://github.com/thisago/whois.nim
```

or

Manually with Nimble

```bash
# Clone repo
git clone https://github.com/thisago/whois.nim

# go to folder
cd whois

# Install (with Nimble)
nimble install -y
```

## Docs

You can find the [docs in Github pages](https://thisago.github.io/whois/whois.html)

## Usage

example/example.nim
```nim
import whois

echo whois("duckduckgo.com")

# or

var domain = "metager.org".toDomain # convert to a `Domain` instance
domain.update() # Get data from API

echo domain
```

Run:
```bash
nim c -r -d:ssl -d:release example/example.nim
```

NOTE: Compile with `-d:ssl`

## TODO

- [ ] Change the nimble package url from gitea to github


## Contributing

Its easy to create a parser for another API!

Just see `src/whois/api_doyosi.nim` to understand how simple is it.

All definitions is in `src/whois/core.nim`.

- Create a new file in `src/whois/` called `api_APINAME.nim`
- Import module `src/whois/core.nim`
- Export a function named `apiFetch`
- In `src/whois.nim`
  - Import the new api parser using:
    ```nim
    from whois/api_APINAME.nim import nil
    ```
  - Call the `apiFetch` function in `update` in end like:
    ```nim
    if api_APINAME.apiFetch(self, noCache): return # If any error, try next API
    ```

## NOTICE

The implemented API service has no relation with this lib.


changelog.md
<!--
  Created at: 06/09/2021 22:46:52 Wednesday
  Modified at: 09/17/2021 12:43:18 AM Friday

        Copyright (C) 2021 Thiago Navarro
  See file "license" for details about copyright
-->


# Changelog

## Version 1.0.2 (09/17/2021 00:43:18 Friday)

- Changed all file headers
- Fixed version in nimble file
- regenerated docs

---

## Version 1.0.1

- Fixed a bug that increments the name servers and not deletes the old ones on update


tests/test1.nim
#[
  Created at: 06/06/2021 23:22:26 Sunday
  Modified at: 09/17/2021 12:41:44 AM Friday

        Copyright (C) 2021 Thiago Navarro
  See file "license" for details about copyright
]#

import unittest

import whois

echo "[NOTICE] The tests will make some requests to API. Depending of your " &
     "internet speed, may take some time"

test "Unavaliable":
  check whois("google.com").avaliable == false
  check whois("bing.com").avaliable == false
  check whois("facebook.com").avaliable == false

test "Avaliable":
  check whois("uknmaiNadd.org").avaliable == true
  check whois("uknmaiNddasde.net").avaliable == true
  check whois("uknaadasdgv.org").avaliable == true

test "Unknown":
  check whois("google.fgff").error == DomainError.unknownAvaliability
  check whois("bing.daf").error == DomainError.unknownAvaliability
  check whois("facebook").error == DomainError.invalidDomain


src/whois/core.nim
#[
  Created at: 06/07/2021 10:38:20 Monday
  Modified at: 09/17/2021 12:42:12 AM Friday

        Copyright (C) 2021 Thiago Navarro
  See file "license" for details about copyright
]#

from std/times import DateTime, utc, parse
from httpcore import HttpCode
from strutils import find, AllChars, Letters, Digits, delete, strip

type
  DomainError* {.pure.} = enum
    ## The possible errors that can have
    none, invalidDomain, unknownAvaliability, apiError

  DomainData* = object
    ## All data provided by API
    id*: string                         ## Registry Domain ID
    creationDate*: DateTime             ## Creation Date
    expireDate*: DateTime               ## Registry Expiry Date
    nameServers*: seq[string]           ## Name Server(s)
    registrantCountry*: string          ## Registrant Country
    registrantOrganization*: string     ## Registrant Organization
    registrantState*: string            ## Registrant State/Province
    registrar*: string                  ## Registrar
    registrarAbuseContactEmail*: string ## Registrar Abuse Contact Email
    registrarAbuseContactPhone*: string ## Registrar Abuse Contact Phone
    registrarIanaId*: string            ## Registrar IANA ID
    updatedDate*: DateTime              ## Updated Date
    whoisDBLastUpdate*: DateTime        ## Last update of WHOIS database
    whoisServer*: string                ## Registrar WHOIS Server

  Domain* = object
    ## The domain object
    name*, tld*: string
    avaliable*: bool
    error*: DomainError
    data*: DomainData
    # cached*: bool # TODO

# Api types
type
  ApiResponse* = object
    ## The response got by API
    code*: HttpCode ## The `Response` http code
    body*: string   ## The `Response` body
    # cached*: bool   ## Inform if was cached TODO

proc full*(self: Domain): string {.noSideEffect.} =
  ## Join the name with tld to get the full domain name
  self.name & "." & self.tld

proc `==`*(a, b: Domain): bool {.noSideEffect.} =
  ## Checks if the full domain name is same
  a.full == b.full

const invalidDateChars = AllChars - Digits - {'T', 'Z', ':', '-'}

proc toDate*(s: string): DateTime =
  let endIndex = s.find invalidDateChars
  var cleaned = s.strip
  if endIndex > -1: cleaned = cleaned.substr(0, endIndex - 1).strip

  if cleaned.len == 0: return

  return parse(cleaned, "yyyy-MM-dd'T'HH:mm:sszzz", utc())


src/whois/api_doyosi.nim
#[
  Created at: 06/07/2021 10:36:49 Monday
  Modified at: 09/17/2021 12:42:19 AM Friday

        Copyright (C) 2021 Thiago Navarro
  See file "license" for details about copyright
]#

##[
  Doyosi free API implementation

  API source code: https://github.com/Karyazilim/DoyosiWhois
]##

{.experimental: "codeReordering".}

from std/httpcore import Http200, `$`
from std/httpclient import newHttpHeaders, newHttpClient, post, code, `==`, body
from std/uri import encodeUrl
from std/json import parseJson, `{}`, getStr
from std/strutils import `%`, toLowerAscii, strip

import std/with
from std/sugar import collect

import kashae

import ./core


const apiConfigs = (
  url: "https://whois.doyosi.com/whois",
  referrer: "https://whois.doyosi.com/",

  # url: "http://127.0.0.1/u3/apache/DoyosiWhois/",
  # referrer: "http://127.0.0.1/u3/apache/DoyosiWhois/",

  bodyTemplate: "domain=$1",
  contentType: "application/x-www-form-urlencoded; charset=UTF-8",
  takenMessage: "Registry Expiry Date: ",
  messageNewline: "<br />\r\n"
)


proc apiFetch*(self: var Domain, noCache = false): bool =
  ## This function fetch the data of domain and returns a `ApiResponse` instance
  ## Returns true on request success
  result = true
  var response = apiRequest(self.full, noCache)
  if response.code != Http200:
    self.error = DomainError.apiError
    return false
  let json = response.body.parseJson

  if json{"status"}.getStr != "success":
    self.error = DomainError.unknownAvaliability
    return false
  self.parse json{"data"}.getStr


proc apiRequest(domain: string): ApiResponse {.cacheOpt: {clearParam}.} =
  var
    headers = newHttpHeaders({
      "content-type": apiConfigs.contentType,
      "referrer": apiConfigs.referrer,
      "x-requested-with": "XMLHttpRequest",
      "accept-language": "en-US,en;q=0.9",
      "accept": "application/json, text/javascript, */*; q=0.01",
    })
    client = newHttpClient(headers = headers)
  let
    body = apiConfigs.bodyTemplate % domain.encodeUrl()
    response = client.post(apiConfigs.url, body)

  result.body = response.body
  result.code = response.code



proc parse(self: var Domain, data: string) =
  ## Parse the API response and saves in the `Domain` instance
  let domain = self.full

  self.avaliable = data.toLowerAscii().find(
    (apiConfigs.takenMessage % domain).toLowerAscii()
  ) == -1

  if self.avaliable: return

  with self.data:
    creationDate = data.get("Creation Date: ").toDate
    expireDate = data.get("Registry Expiry Date: ").toDate
    id = data.get("Registry Domain ID: ")
    registrantCountry = data.get("Registry Domain ID: ")
    registrantOrganization = data.get("Registrant Organization: ")
    registrantState = data.get("Registrant State/Province: ")
    registrar = data.get("Registrar: ")
    registrarAbuseContactEmail = data.get("Registrar Abuse Contact Email: ")
    registrarAbuseContactPhone = data.get("Registrar Abuse Contact Phone: ")
    registrarIanaId = data.get("Registrar IANA ID: ")
    updatedDate = data.get("Updated Date: ").toDate
    whoisDBLastUpdate = data.get("Last update of whois database: ").toDate
    whoisServer = data.get("Registrar WHOIS Server: ")

  self.data.nameServers = @[]

  block:
    const toFind = "Name Server: " # The value name to get
    var
      mdata = data # A mutable copy of data
      index = mdata.find toFind
    while index > -1: # While finding the value
      mdata = mdata.substr(index + toFind.len).strip # delete everything from
                                                      # begin of match and save
                                                      # on itself
      index = mdata.find toFind # update the index

      self.data.nameServers.add mdata.substr( # Delete all from newline
        0, mdata.find(apiConfigs.messageNewline) - 1)


proc get(data, val: string): string =
  ## Gets the data from response
  let index = data.find val
  if index == -1: return

  let croppedStart = data.substr(index + val.len).strip
  result = croppedStart.substr(
    0, croppedStart.find(apiConfigs.messageNewline) - 1)

proc find(s, sub: string): int =
  strutils.find(s.toLowerAscii(), sub.toLowerAscii())


src/whois.nim
#[
  Created at: 06/06/2021 23:24:50 Sunday
  Modified at: 09/17/2021 12:42:37 AM Friday

        Copyright (C) 2021 Thiago Navarro
  See file "license" for details about copyright
]#


from std/strutils import split, strip

import whois/core; export core

# API selection
from whois/api_doyosi import nil


proc toDomain*(fullDomain: string): Domain {.noSideEffect.} =
  ## Convert a full domain string in a `Domain` instance
  let splitted = fullDomain.split "."
  if splitted.len != 2:
    result.error = DomainError.invalidDomain
    return
  result.name = splitted[0]
  result.tld = splitted[1]


proc update*(self: var Domain, noCache = false) =
  ## Updates the domain with api data
  if self.error != DomainError.none: return

  if api_doyosi.apiFetch(self, noCache): return # If any error, try next API


proc whois*(domain: string, noCache = false): Domain {.inline.} =
  ## A alias for:
  ## .. code-block:: nim
  ##   var domain = "duckduckgo.com".toDomain
  ##   domain.update()
  result = domain.toDomain()
  result.update(noCache)


when isMainModule:
  import times

  echo cputime()

  var domain = "google.co".toDomain
  var a: DomainData

  echo domain

  domain.update()

  echo cputime()
  echo domain

  echo "update again"

  domain.update()

  echo cputime()
  echo domain

  echo "update again"

  domain.update()

  echo cputime()
  echo domain

  echo "update again"

  domain.update(noCache = true)

  echo cputime()
  echo domain
